### ISSUE TITLE
Segfault on merge in std::string::compare and DB::MergeTreeDataPart::remove()

### ISSUE THREAD

**Issue by blind-oracle** [yandex/ClickHouse#5543]
**Describe the bug**
CH segfaults from time to time on inserts.
I'm using https://github.com/lomik/carbon-clickhouse to ingest data.

**How to reproduce**
* 19.6.2.1 and 19.5.4.22 confirmed (build from Altinity and self-built checked)
* HTTP

```
CREATE TABLE IF NOT EXISTS default.graphite (
  Path String CODEC(ZSTD(2)),
  Value Float64 CODEC(Delta, ZSTD(2)),
  Time UInt32 CODEC(Delta, ZSTD(2)),
  Date Date CODEC(Delta, ZSTD(2)),
  Timestamp UInt32 CODEC(Delta, ZSTD(2))
) ENGINE = ReplicatedGraphiteMergeTree('/clickhouse/tables/01/graphite', 'hostname', 'graphite_rollup')
PARTITION BY toYYYYMM(Date)
ORDER BY (Path, Time);
```

**Error message and/or stacktrace**
```
BaseDaemon: ########################################
BaseDaemon: (version 19.6.2.1) (from thread 37) Received signal Segmentation fault (11).
BaseDaemon: Address: 0xffffffffffffffe8
BaseDaemon: Access: read.
BaseDaemon: Address not mapped to object.
BaseDaemon: 0. clickhouse-server(std::string::compare(std::string const&) const+0x11) [0x319fda1]
BaseDaemon: 1. clickhouse-server(DB::MergeTreePartInfo::operator<(DB::MergeTreePartInfo const&) const+0x12) [0x6152c32]
BaseDaemon: 2. clickhouse-server(DB::MergeTreeData::modifyPartState(boost::multi_index::detail::bidir_node_iterator<boost::multi_index::detail::ordered_index_node<boost::multi_index::detail::null_augment_policy, boost::multi_index::detail::index_node_base<std::shared_ptr<DB::MergeTreeDataPart const>, std::allocator<std::shared_ptr<DB::MergeTreeDataPart const> > > > >, DB::MergeTreeDataPart::State)+0x74) [0x61f7494]
BaseDaemon: 3. clickhouse-server(DB::MergeTreeData::grabOldParts()+0x1ab) [0x61e541b]
BaseDaemon: 4. clickhouse-server(DB::StorageReplicatedMergeTree::clearOldPartsAndRemoveFromZK()+0x59) [0x6179409]
BaseDaemon: 5. clickhouse-server(DB::ReplicatedMergeTreeCleanupThread::iterate()+0x16) [0x627e036]
BaseDaemon: 6. clickhouse-server(DB::ReplicatedMergeTreeCleanupThread::run()+0xcb) [0x627e27b]
BaseDaemon: 7. clickhouse-server(DB::BackgroundSchedulePoolTaskInfo::execute()+0xe1) [0x5ddc851]
BaseDaemon: 8. clickhouse-server(DB::BackgroundSchedulePool::threadFunction()+0x62) [0x5ddcf22]
BaseDaemon: 9. clickhouse-server() [0x5ddcf84]
BaseDaemon: 10. clickhouse-server(ThreadPoolImpl<std::thread>::worker(std::_List_iterator<std::thread>)+0x187) [0x68d7947]
BaseDaemon: 11. clickhouse-server() [0x745371f]
BaseDaemon: 12. /lib64/libpthread.so.0(+0x7dd5) [0x7fdbdd563dd5]
BaseDaemon: ########################################
```

```
BaseDaemon: ########################################
BaseDaemon: (version 19.6.2.1) (from thread 30) Received signal Segmentation fault (11).
BaseDaemon: Address: 0x7f9bffffffe8
BaseDaemon: Access: read.
BaseDaemon: Address not mapped to object.
BaseDaemon: 0. clickhouse-server(DB::MergeTreeDataPart::remove() const+0x16) [0x620a3f6]
BaseDaemon: 1. clickhouse-server() [0x615fc88]
BaseDaemon: 2. clickhouse-server(DB::StorageReplicatedMergeTree::clearOldPartsAndRemoveFromZK()+0x1e2) [0x6179592]
BaseDaemon: 3. clickhouse-server(DB::ReplicatedMergeTreeCleanupThread::iterate()+0x16) [0x627e036]
BaseDaemon: 4. clickhouse-server(DB::ReplicatedMergeTreeCleanupThread::run()+0xcb) [0x627e27b]
BaseDaemon: 5. clickhouse-server(DB::BackgroundSchedulePoolTaskInfo::execute()+0xe1) [0x5ddc851]
BaseDaemon: 6. clickhouse-server(DB::BackgroundSchedulePool::threadFunction()+0x62) [0x5ddcf22]
BaseDaemon: 7. clickhouse-server() [0x5ddcf84]
BaseDaemon: 8. clickhouse-server(ThreadPoolImpl<std::thread>::worker(std::_List_iterator<std::thread>)+0x187) [0x68d7947]
BaseDaemon: 9. clickhouse-server() [0x745371f]
BaseDaemon: 10. /lib64/libpthread.so.0(+0x7dd5) [0x7f9d314d5dd5]
BaseDaemon: (version 19.6.2.1) (from thread 47) Terminate called without an active exception
BaseDaemon: ########################################
```

**Comment by alexey-milovidov** [comment]
> Date Date CODEC(Delta

The issue was fixed in #5786
(the fix is available in 19.11 testing)

**Comment by blind-oracle** [comment]
Thanks! The fix also went to 19.x stable release.
